macaw-riscv-symbolic/tests/Main.hs

```diff
@@ -0,0 +1,218 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE ImplicitParams #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+module Main (main) where
+
+import           Control.Lens ( (^.) )
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Char8 as BS8
+import qualified Data.ElfEdit as Elf
+import qualified Data.Foldable as F
+import qualified Data.Map as Map
+import           Data.Maybe ( mapMaybe )
+import qualified Data.Parameterized.Classes as PC
+import qualified Data.Parameterized.Nonce as PN
+import           Data.Parameterized.Some ( Some(..) )
+import           Data.Proxy ( Proxy(..) )
+import           GHC.TypeNats ( KnownNat, type (<=) )
+import qualified Prettyprinter as PP
+import           System.FilePath ( (</>), (<.>) )
+import qualified System.FilePath.Glob as SFG
+import qualified System.IO as IO
+import qualified Test.Tasty as TT
+import qualified Test.Tasty.HUnit as TTH
+import qualified Test.Tasty.Options as TTO
+import qualified Test.Tasty.Runners as TTR
+
+import qualified Data.Macaw.Architecture.Info as MAI
+import qualified Data.Macaw.CFG as MC
+import qualified Data.Macaw.Discovery as M
+import qualified Data.Macaw.Memory.ElfLoader.PLTStubs as MMELP
+import qualified Data.Macaw.Symbolic as MS
+import qualified Data.Macaw.Symbolic.Testing as MST
+import qualified Data.Macaw.RISCV as MR
+import           Data.Macaw.RISCV.Symbolic ()
+import qualified GRIFT.Types as G
+import qualified What4.Config as WC
+import qualified What4.Expr.Builder as WEB
+import qualified What4.Interface as WI
+import qualified What4.ProblemFeatures as WPF
+import qualified What4.Solver as WS
+
+import qualified Lang.Crucible.Backend as CB
+import qualified Lang.Crucible.Backend.Online as CBO
+import qualified Lang.Crucible.Simulator as CS
+import qualified Lang.Crucible.LLVM.MemModel as LLVM
+
+-- | A Tasty option to tell us to save SMT queries and responses to /tmp for debugging purposes
+data SaveSMT = SaveSMT Bool
+  deriving (Eq, Show)
+
+instance TTO.IsOption SaveSMT where
+  defaultValue = SaveSMT False
+  parseValue v = SaveSMT <$> TTO.safeReadBool v
+  optionName = pure "save-smt"
+  optionHelp = pure "Save SMT sessions to files in /tmp for debugging"
+
+-- | A tasty option to have the test suite save the macaw IR for each test case to /tmp for debugging purposes
```

Comment:
I'm wondering if these should just save to the current directory rather than `/tmp`.  Dropping files in a global location mixes them with other runs and could lead to looking at the wrong file.  Another thought is that if we want to generate these during CI and possibly save them as artifacts, it will be easier to generate them directly in the right location (e.g. `mkdir test_out; cd test_out; cabal run macaw-riscv-symbolic-tests -- --save-smt --save-macaw` with the `test_out` being designated in the `ci.yaml` `artifacts` clause).

Reply:
This is another piece of code that I copied over from the `macaw-*-symbolic` test suites. I don't have a strong feeling one way or another about where these files live. Given that `--save-smt` isn't enabled by default, I think we could put these files anywhere.

In fact, I think it would be better if these options took an explicit directory as an argument so that the user can control where the files should live. That way, we could just say `cabal run macaw-riscv-symbolic-tests --save-smt test_out --save-macaw test_out` directly, without needing to `cd` to and fro.

Does that sound reasonable? If so, I'll open a separate issue to track this idea.

Reply:
That sounds reasonable, and that way we can address that concern independently and simultaneously for all similar situations.

Reply:
I've opened https://github.com/GaloisInc/macaw/issues/413 for this.